pkg/loghttp/push/push.go

```diff
@@ -217,16 +227,33 @@ func ParseLokiRequest(userID string, r *http.Request, tenantsRetention TenantsRe
 	pushStats.ContentType = contentType
 	pushStats.ContentEncoding = contentEncoding
 
-	for _, s := range req.Streams {
+	discoverServiceName := limits.DiscoverServiceName(userID)
+	for i := range req.Streams {
+		s := req.Streams[i]
 		pushStats.StreamLabelsSize += int64(len(s.Labels))
 
-		var lbs labels.Labels
-		if tenantsRetention != nil || tracker != nil {
-			lbs, err = syntax.ParseLabels(s.Labels)
-			if err != nil {
-				return nil, nil, fmt.Errorf("couldn't parse labels: %w", err)
+		lbs, err := syntax.ParseLabels(s.Labels)
+		if err != nil {
+			return nil, nil, fmt.Errorf("couldn't parse labels: %w", err)
+		}
+
+		if !lbs.Has(LabelServiceName) && len(discoverServiceName) > 0 {
+			serviceName := ServiceUnknown
+			for _, labelName := range discoverServiceName {
+				if labelVal := lbs.Get(labelName); labelVal != "" {
+					serviceName = labelVal
+					break
+				}
 			}
+
+			lb := labels.NewBuilder(lbs)
+			lbs = lb.Set(LabelServiceName, serviceName).Labels()
+			s.Labels = lbs.String()
+
+			// Remove the added label after it's added to the stream so it's not consumed by subsequent steps
```

Comment:
I wonder if this is the right call here :thinking: It's consistent with the current behavior here, but I think it might be a bit confusing to users. They will suddenly see `service_name` everywhere, so they might wonder why they couldn't configure their retention periods with it too, right? :thinking:

Reply:
I agree that this is kind of confusing but I believe the tenant retentions bytes here are used for billing so I'm a bit cautious to change it.

Also we've not made anything worse 😅